nimbus/src/enrollment.rs

```diff
@@ -427,12 +435,34 @@ impl EnrollmentStatus {
             enrollment_id: Uuid::new_v4(),
         }
     }
+
     // This is used in examples, but not in the main dylib, and
     // triggers a dead code warning when building with `--release`.
     #[allow(dead_code)]
     pub fn is_enrolled(&self) -> bool {
         matches!(self, EnrollmentStatus::Enrolled { .. })
     }
+
+    /// Make a clone of this status, but with the enrollment_id to a new random value.
+    fn clone_with_new_enrollment_id(&self) -> Self {
+        let mut updated = self.clone();
+        match updated {
+            EnrollmentStatus::Enrolled {
+                ref mut enrollment_id,
+                ..
+            }
+            | EnrollmentStatus::Disqualified {
+                ref mut enrollment_id,
+                ..
+            }
+            | EnrollmentStatus::WasEnrolled {
+                ref mut enrollment_id,
+                ..
+            } => *enrollment_id = Uuid::new_v4(),
+            EnrollmentStatus::NotEnrolled { .. } | EnrollmentStatus::Error { .. } => (),
+        };
+        updated
+    }
```

Comment:
Do we want to keep the current enrollment ids?  Since we are only keeping the experiment enrollments to prevent some changes from running away into the app, and we aren't sending any of this in telemetry, I don't know that we need to worry about rotating the enrollment_ids here.  If the client re-enabled telemetry, some of these enrollments might get weird if the randomization unit changes and they are no longer in a targeted bucket.

Reply:
> Do we want to keep the current enrollment ids?

I don't think we should. The case I'm worried about here is something like:

1. User gets enrolled in an experiment, emits a telemetry `Enrolled` event with `enrollment_id=AAA`.
2. User disables telemetry.
3. User re-enables telemetry, remaining enrolled with the same enrollment id.
4. The experiment ends, emitting a telemetry `Unenrolled` event with `enrollment_id=AAA`.

We could use the events from (1) and (4) to link the user's telemetry from before disablement with their telemetry after re-enablement.

Reply:
(OTOH, in theory we're supposed to have deleted the user's telemetry from before disablement, so maybe that's not as important as it might seem...?)

Reply:
As a point of comparison, [the desktop implementation](https://searchfox.org/mozilla-central/rev/8883276967d39918e2ce64e873afdd432fb406ca/toolkit/components/normandy/lib/PreferenceExperiments.jsm#387) sets enrollment ids to a canary value of `__NO_ENROLLMENT_ID__`.

Reply:
You are right, if we kept the same enrollments and ids when re-enabling telemetry, if we sent out any events, it would probably look like a new user (because the telemetry id was rotated) that had an unenrollment (or disqualification) event without a matching enrollment event. So, resetting the ids to a canary value I would definitely approve of (we use a `c0ffee` uuid in Telemetry for this).  

I would really prefer a way to just delete all the enrollment info just like we delete all the telemetry associated with a client, but I do understand the concern about what that might mean to the application if we can't guarantee the same result to multiple calls of getBranch, etc. Maybe this is a good question to bring up with the product team(s) to see what sort of behavior they would expect in this case? I don't know that this really needs to be addressed immediately, rather a good to think about as a possible iteration.